### PR TITLE
LibWeb: Implement `scroll_height` for Element

### PR DESCRIPTION
**Pull request by guerinoni**

### CODE REVIEW

**Review by linusg — CHANGES_REQUESTED** [review]
Implemenet -> Implement

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 1]
```diff
@@ -1105,8 +1105,33 @@ int Element::scroll_width() const
...
+    auto viewport_height = document.browsing_context()->viewport_rect().height().value();
+    if (paint_box()) {
+        viewport_height -= paint_box()->absolute_padding_box_rect().height().value();
+    }
+
+    // 4. If the element is the root element and document is not in quirks mode return max(viewport scrolling area height, viewport height).
+    if (document.document_element() == this && document.in_quirks_mode())
+        return AK::max(document.browsing_context()->viewport_rect().height().value(), viewport_height);
+
+    //    // 5. If the element is the body element, document is in quirks mode and the element is not potentially scrollable, return max(viewport scrolling area height, viewport height).
```
[the remark this answers is not in the document]
Comment (guerinoni) [reply Userland/Libraries/LibWeb/DOM/Element.cpp]: Sorry :) fixed

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 2]
```diff
@@ -1105,8 +1105,33 @@ int Element::scroll_width() const
...
+
+    // 5. If the element is the body element, document is in quirks mode and the element is not potentially scrollable, return max(viewport scrolling area height, viewport height).
+    if (document.body() == this && document.in_quirks_mode() && !is_potentially_scrollable())
+        return AK::max(document.browsing_context()->viewport_rect().height().value(), viewport_height);
+
+    // 6. If the element does not have any associated box return zero and terminate these steps.
+    if (!layout_node() || !is<Layout::Box>(layout_node()))
+        return 0;
+
+    // 7. TODO: Return the height of the element’s scrolling area.
```
Comment (GMTA) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1133]: Maybe have one comment mirroring the spec text and a separate `// FIXME: ...` explaining what is wrong and what is actually should be doing.

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 3]
```diff
@@ -1105,8 +1105,33 @@ int Element::scroll_width() const
```
Comment (davidot) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1105]: Could you add a spec link? (https://w3c.github.io/csswg-drafts/cssom-view/#dom-element-scrollheight I think)

**Review by davidot — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/DOM/Element.cpp (hunk 3 above)
Comment (guerinoni) [reply Userland/Libraries/LibWeb/DOM/Element.cpp:1105]: Sure

File: Userland/Libraries/LibWeb/DOM/Element.cpp (hunk 2 above)
Comment (guerinoni) [reply Userland/Libraries/LibWeb/DOM/Element.cpp]: It was a leftover :)

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 4]
```diff
@@ -1105,8 +1105,33 @@ int Element::scroll_width() const
...
+        return 0;
+
+    // 3. Let viewport height be the height of the viewport excluding the height of the scroll bar, if any, or zero if there is no viewport.
+    auto viewport_height = document.browsing_context()->viewport_rect().height().value();
+    if (paint_box()) {
+        viewport_height -= paint_box()->absolute_padding_box_rect().height().value();
+    }
+
+    // 4. If the element is the root element and document is not in quirks mode return max(viewport scrolling area height, viewport height).
+    if (document.document_element() == this && document.in_quirks_mode())
```
[the remark this answers is not in the document]
Comment (guerinoni) [reply Userland/Libraries/LibWeb/DOM/Element.cpp]: Fixed thanks!!

File: Userland/Libraries/LibWeb/DOM/Element.cpp (hunk 3 above)
Comment (davidot) [reply Userland/Libraries/LibWeb/DOM/Element.cpp:1105]: Thanks!

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 5]
```diff
@@ -1103,10 +1103,36 @@ int Element::scroll_width() const
...
+    // 5. If the element is the body element, document is in quirks mode and the element is not potentially scrollable, return max(viewport scrolling area height, viewport height).
+    if (document.body() == this && document.in_quirks_mode() && !is_potentially_scrollable())
+        return AK::max(document.browsing_context()->viewport_rect().height().value(), viewport_height);
+
+    // 6. If the element does not have any associated box return zero and terminate these steps.
+    if (!layout_node() || !is<Layout::Box>(layout_node()))
+        return 0;
+
+    // 7. Return the height of the element’s scrolling area.
+    return document.browsing_context()->viewport_rect().height().value();
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1135]: That's not the height of the element's scrolling area, that's just the viewport height :)

The scrolling area is defined [here](https://w3c.github.io/csswg-drafts/cssom-view/#scrolling-area) and basically requires calculating the overflow of the element's corresponding layout box.

Comment (guerinoni) [reply Userland/Libraries/LibWeb/DOM/Element.cpp:1135]: Thanks for feedback, I don't have a clear mind of what I need to calculate 😓 can you explain more in details? 😅

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 6]
```diff
@@ -1122,9 +1122,36 @@ int Element::scroll_width() const
...
+    auto& document = this->document();
+
+    // 2. If document is not the active document, return zero and terminate these steps.
+    if (!document.is_active())
+        return 0;
+
+    // 3. Let viewport height be the height of the viewport excluding the height of the scroll bar, if any, or zero if there is no viewport.
+    auto viewport_height = document.browsing_context()->viewport_rect().height().value();
+    if (paint_box()) {
+        viewport_height -= paint_box()->absolute_padding_box_rect().height().value();
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1138]: I don't understand how the padding box height is related to the height of the scroll bar.

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 7]
```diff
@@ -1122,9 +1122,36 @@ int Element::scroll_width() const
...
+    }
+
+    // 4. If the element is the root element and document is not in quirks mode return max(viewport scrolling area height, viewport height).
+    if (document.document_element() == this && !document.in_quirks_mode())
+        return AK::max(document.browsing_context()->viewport_rect().height().value(), viewport_height);
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1143]: The height of the viewport rect is not the height of the viewport scrolling area. My understanding is, the "scrolling area" is the size of the content inside the viewport. See the "rightward and downward"/"For a viewport" box in this table: https://w3c.github.io/csswg-drafts/cssom-view/#scrolling-area

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 8]
```diff
@@ -1122,9 +1122,36 @@ int Element::scroll_width() const
...
+
+    // 5. If the element is the body element, document is in quirks mode and the element is not potentially scrollable, return max(viewport scrolling area height, viewport height).
+    if (document.body() == this && document.in_quirks_mode() && !is_potentially_scrollable())
+        return AK::max(document.browsing_context()->viewport_rect().height().value(), viewport_height);
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1147]: Same comment about the scrolling area height.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]

**Pull request closed without merging**